### PR TITLE
Add docs comments and tests for TypeScript definitions

### LINKED ISSUE #293: Add documentation comments and tests for the TypeScript definition
To improve the TS definition.

See https://github.com/sindresorhus/typescript-definition-style-guide on how to do it.

We actually already [have TS tests](https://github.com/chalk/chalk/blob/master/types/test.ts), but they should be converted to use `tsd-check`.

We should also rename `ChalkOptions` to `Options` and `ChalkConstructor` to `Constructor` with type aliases to preserve backwards compatibility.

### PR DESCRIPTION
**Pull request by GeorgeGkas**
Fixes #293.

TODO:
- [x] Add [tsdoc](https://github.com/Microsoft/tsdoc) style comments to TS definition file.
  - [x] Should document the following modifier properties of `Chalk` interface: `hidden`, `strikethrough` and `visible`. (Could not find the appropriate words to do it. Any help is welcome!)
- [x] Rename `ChalkOptions` to `Options`.
- [x] Type alias `Options` to `ChalkOptions` to preserve backwards compatibility.
- [x] Rename `ChalkConstructor` to `Constructor`.
- [x] Type alias `Constructor` to `ChalkConstructor` to preserve backwards compatibility.
- [x]  Add [tsd-check](https://github.com/SamVerschueren/tsd-check#readme) module to test the TS definitions. 
- [x] Write tests for TS definitions using [tsd-check](https://github.com/SamVerschueren/tsd-check#readme).

Reviews are welcome! :smiley:

### CODE REVIEW

**Comment by GeorgeGkas** [comment]
`tsd-check` does not support Node v4 as it use ES6 destructing assignment in its core. Travis job on Node v4 [fails](https://travis-ci.org/chalk/chalk/jobs/429568924#L714) due to this reason (builds on Node v6 and Node 8 succeed).

Should we drop support for Node v4? We should certainly update `.travis.yml`.

**Comment by Qix-** [comment]
> Should we drop support for Node v4?

@sindresorhus I vote yes, though not sure what the release would look like. That's unrelated to this PR though, `meow` isn't a production dependency.

**Comment by sindresorhus** [comment]
I have dropped support for Node.js 4 in master. Please rebase from master.

File: package.json [hunk 1]
```diff
@@ -64,7 +65,8 @@
 			"mocha"
 		],
 		"ignores": [
-			"test/_flow.js"
+			"test/_flow.js",
+			"*.ts"
```
Comment (sindresorhus) [inline package.json]: This is not required

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+ * Order doesn't matter, and later styles take precedent in case of a conflict.
+ * This simply means that `chalk.red.yellow.green` is equivalent to `chalk.green`.
+ */
+declare const chalk: Chalk & { supportsColor: ColorSupport };
+
+/**
+ * Alias Options and Constructor to preserve backwards compatibility.
+ */
+export type ChalkOptions = Options;
+export type ChalkConstructor = Constructor;
```
Comment (sindresorhus) [inline index.d.ts]: We're doing a major release now, so this is not needed after all.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+export default chalk
```
Comment (sindresorhus) [inline index.d.ts]: Semi-colon

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+	hwb(h: number, w: number, b: number): this;
+
+	/**
+	 * Use HEX value to set background color.
+	 *
+	 * @param color - Hexadecimal value representing the desired color.
+	 *
+	 * @example
+	 *
+	 * import chalk from 'chalk';
```
Comment (sindresorhus) [inline index.d.ts]: Typo

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	keyword(color: string): this;
+
+	/**
+	 * Use RGB values to set text color.
+	 *
+	 * @param r - red
+	 * @param g - green
+	 * @param b - blue
+	 */
+	rgb(r: number, g: number, b: number): this;
```
Comment (sindresorhus) [inline index.d.ts]: Use the full words, not `r` and `g`. Same with the other methods. The params then don't need any `@param` docs.

File: index.d.ts (hunk 4 above)
Comment (GeorgeGkas) [reply index.d.ts]: what is the typo?

**Comment by GeorgeGkas** [comment]
just did the rebase. I'm start applying review changes now.

**Comment by GeorgeGkas** [comment]
the last rebase introduced two eslint errors. I am going to push another commit to fix [these](https://travis-ci.org/chalk/chalk/builds/430008320?utm_source=github_status&utm_medium=notification).

**Comment by sindresorhus** [comment]
> Should document the following modifier properties of Chalk interface: hidden, strikethrough and visible. (Could not find the appropriate words to do it. Any help is welcome!)

- `strikethrough`: https://en.wikipedia.org/wiki/Strikethrough
- `hidden`: Prints the text, but makes it invisible.
- `visible`: Prints the text only when Chalk is enabled. Can be useful for things that are purely cosmetic.

Would be great if you could update the docs for the modifiers (modifiers only, not the colors) in the readme with the same docs as in the TS types.

**Comment by GeorgeGkas** [comment]
The last commit has stuck in pending status in Travis CI. Do you mind to force restart it? I do not know what happened.

**Comment by GeorgeGkas** [comment]
done

File: readme.md [hunk 6]
```diff
@@ -166,15 +166,17 @@ Explicit 256/Truecolor mode can be enabled using the `--color=256` and `--color=
...
+| --- | --- |
+| `reset` | Resets the current color chain. |
+| `bold` | Make text bold. |
+| `dim` | Emitting only a small amount of light. |
+| `italic` | Make text italic. *(Not widely supported)* |
+| `underline` | Make text underline. *(Not widely supported)* |
+| `inverse` | Inverse background and foreground colors. |
+| `hidden` | Prints the text, but makes it invisible. |
+| `strikethrough` | [https://en.wikipedia.org/wiki/Strikethrough](https://en.wikipedia.org/wiki/Strikethrough) |
+| `visible` | Prints the text only when Chalk is enabled. Can be useful for things that are purely cosmetic. |
```
Comment (sindresorhus) [inline readme.md]: Can we not use tables here? They're annoying to maintain and doesn't give much benefit.

You can just use the format:

- `reset` - Resets the current color chain.

File: index.d.ts [hunk 7]
```diff
@@ -225,17 +225,17 @@ export interface Chalk {
...
 
 	/**
-	 * Modifier
+	 * Modifier: Prints the text, but makes it invisible.
 	 */
 	readonly hidden: this;
 
 	/**
-	 * Modifier
+	 * Modifier: https://en.wikipedia.org/wiki/Strikethrough
```
Comment (sindresorhus) [inline index.d.ts]: Sorry, I was unclear. I didn't mean to use this literally, but rather to find a good description from the Wikipedia page.

> Puts a horizontal line through the center of the text.

It should also contain the `(Not widely supported)` part from the readme.

**Comment by sindresorhus** [comment]
Ping :)

**Comment by GeorgeGkas** [comment]
@sindresorhus sorry for the delay. I was busy lately. I'm working on this right now.

**Comment by sindresorhus** [comment]
Looks good now. Thank you 🙌

**Pull request merged**